### ISSUE TITLE
L2 & learn_many

### ISSUE THREAD

**Issue by ColdTeapot273K** [online-ml/river#881]
Hey, I've noticed L2 hyperparameter is no-op when computing objective functions for linear models via `learn_many`:
https://github.com/online-ml/river/blob/main/river/linear_model/glm.py#L106-L123

I've tried implementing it like so:
https://github.com/ColdTeapot273K/river/commit/a7cfffd0c375efc7eae3c85c3160821187d31b73

But my tests on synth data with high amount of non-informative features show that it reliably worsens the metrics:

https://gist.github.com/ColdTeapot273K/471dba785fb80012d6921fbddc482d0a
 
@MaxHalford got any ideas what am I missing?

**Comment by MaxHalford** [comment]
Hey there, good catch.

Have you checked if your implementation produces the same results as scikit-learn's `SGDClassifier` and `SGDRegressor`? I would start there.

**Comment by ColdTeapot273K** [comment]
@MaxHalford 
I've checked here:

https://gist.github.com/ColdTeapot273K/471dba785fb80012d6921fbddc482d0a

and w/ a couple of other similar tests

The results are not precisely the same, but conclusion is that given the data, in `sklearn` `SGDRegressor` higher `l2` practically reduces the loss (as intended) and in my impl for `river` it's the opposite (not intended).

I've tried digging into sklearn sources for `SGDRegressor` and oh boy are they obfuscated - some kind of scaling everywhere:
https://github.com/scikit-learn/scikit-learn/blob/37ac6788c9504ee409b75e5e24ff7d86c90c2ffb/sklearn/linear_model/_sgd_fast.pyx#L599-L605

https://github.com/scikit-learn/scikit-learn/blob/37ac6788c9504ee409b75e5e24ff7d86c90c2ffb/sklearn/utils/_weight_vector.pyx.tp

**Comment by MaxHalford** [comment]
I see! I guess the implementation needs more digging into. I recall that River and scikit-learn produce exactly the same results in the single instance case. I think we even have unit tests for that.

**Comment by Joseph-iris** [comment]
Hey @MaxHalford @ColdTeapot273K  my team noticed the same issue, following this as well

**Comment by MaxHalford** [comment]
I'll take a look soon :)

**Comment by MaxHalford** [comment]
Ok I just implemented L2 regularization for `learn_many`. I believe your implementation was correct @ColdTeapot273K. A subtle difference is that you need to change the default squared loss to match with scikit-learn's convention. I suggest checking out the `test_glm.py` file for examples.

River and scikit-learn will necessarily produce different results because internally scikit-learn doesn't do batching. Indeed it actually loops over the samples one by one. However I reran your example with a learning rate of 0.01 instead of 0.1 and the results were of the same order of magnitude.

**Comment by ColdTeapot273K** [comment]
Thanks for helping @MaxHalford

**Comment by hoanganhngo610** [comment]
Since L2 (and also L1 regularization has been implemented by @ColdTeapot273K ) is working properly now, I would proceed to close this issue!
